[RFC] Allow packed types to transitively contain aligned types (#3718)
rust-lang/rfcs: text/0000-layout-packed-aligned.md

- Start Date: 2024-10-24
- RFC PR: [rust-lang/rfcs#3718](https://github.com/rust-lang/rfcs/pull/3718)
- Rust Issue: [rust-lang/rust#100743](https://github.com/rust-lang/rust/issues/100743)

# Summary
[summary]: #summary

This RFC makes it legal to have `#[repr(C)]` structs that are:
- Both packed and aligned.
- Packed, and transitively contains`#[repr(align)]` types.

It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.
It has the same behavior as `#[repr(C)]` except on `*-pc-windows-gnu` targets where it uses the msvc layout
rules instead.

# Motivation
[motivation]: #motivation

This RFC enables the following struct definitions:

```rs
#[repr(C, packed(2), align(4))]
struct Foo { // Alignment = 4, Size = 8
    a: u8,   // Offset = 0
    b: u32,  // Offset = 2
}
```

This is commonly needed when Rust is being used to interop with existing C and C++ code bases, which may contain
unaligned types. For example in `clang` it is possible to create the following type definition, and there is
currently no easy way to create a matching Rust type:

```cpp
struct  __attribute__((packed, aligned(4))) MyStruct {
    uint8_t a;
    uint32_t b;

> };

Author: This is the correct behavior, yes.

```

> Currently `#[repr(packed(_))]` structs cannot transitively contain #[repr(align(_))] structs due to differing behavior between msvc and gcc/clang.

Author: Thanks. I've linked this.

However, in most cases, the user would expect `#[repr(C)]` to produce a struct layout matching the same type as defined by the current target. 

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## `#[repr(C)]`

> When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,

Author: Good call. Here's what I added:

## `#[repr(Rust)]`
When `align(N)` and `packed(M)` attributes exist on the same type, or when `packed` structs contain `aligned` fields, the type will have their base alignment increased to `N`, while the struct fields will be laid out as if their
alignments were decreased to `M`. However, in general Rust is free to reorder these fields for optimization purposes, and the only guarantee is that the fields will maintain a minimum alignment of `M`.

> When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,

Reviewer: Since you've added the section, resolving this thread

> the resulting layout matches the target toolchain ABI.

Author: I do believe so.

https://clang.llvm.org/docs/MSVCCompatibility.html

> First, Clang attempts to be ABI-compatible, meaning that Clang-compiled code should be able to link against MSVC-compiled code successfully.

> Record layout: Complete. We’ve tested this with a fuzzer and have fixed all known bugs.

For example, given:
```c
#[repr(C, align(4))]
struct Foo(u8);
#[repr(C, packed(1))]
struct Bar(Foo);
```
`align_of::<Bar>()` would be 4 for `*-pc-windows-msvc` and 1 for everything else.

## `#[repr(system)]`
When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,
the resulting layout matches the target OS ABI.

For example, given:
```c
#[repr(C, align(4))]
struct Foo(u8);
#[repr(C, packed(1))]
struct Bar(Foo);
```

> `align_of::<Bar>()` would be 4 for `*-pc-windows-msvc` and `*-pc-windows-gnu`. It would be 1 for everything else.

Reviewer 2: Are the structs in these examples supposed to be `#[repr(system, ...)]`

Author: Good catch, thanks!

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

In the following paragraphs, "Decreasing M to N" means:
```
if M > N {
    M = n
}
```

"Increasing M to N" means:
```
if M < N {
    M = N
}
```

`#[repr(align(N))]` increases the base alignment of a type to be N.

`#[repr(packed(M))]` decreases the alignment of the struct fields to be M. Because the base alignment of the type
is defined as the maximum of the alignment for any fields, this also has the indirect result of decreasing the base
alignment of the type to be M.

When the align and packed modifiers are applied on the same type as `#[repr(align(N), packed(M))]`,
the alignment of the struct fields are decreased to be M. Then, the base alignment of the type is
increased to be N.

> When a `#[repr(packed(M))]` struct transitively contains a field with `#[repr(align(N))]` type,

Author: Thanks, updated.

> - The field is first `pad_to_align`. Then, the field is added to the struct with alignment decreased to M. The packing requirement overrides the alignment requirement. (GCC, `#[repr(Rust)]`, `#[repr(C)]` on gnu targets, `#[repr(system)]` on non-windows targets)

Author: However, Clang and GCC has the following layout:

```c
struct  __attribute__((aligned(4)))
MyStructInner { // Alignment = 4, Size = 4
    uint8_t a; // Offset = 0
};

struct  __attribute__((packed))
MyStruct { // Alignment = 1, Size = 8
    MyStructInner a; // Offset = 0
    uint32_t b; // Offset = 4
};
```

It is indeed curious that `b` would have offset = 4 in `MyStruct`, even though it's not technically necessary. This behavior was covered by the language I used in the RFC.

> - The field is first `pad_to_align`. Then, the field is added to the struct with alignment decreased to M. The packing requirement overrides the alignment requirement. (GCC, `#[repr(Rust)]`, `#[repr(C)]` on gnu targets, `#[repr(system)]` on non-windows targets)

Reviewer: No, there's no curiosity here? `MyStructInner` has a size of 4, so the next field to follow it must be offset by at least 4 bytes from the field of type `MyStructInner` to avoid overlapping it. It's not allowed to put fields within the trailing padding of other fields in C, C++, nor Rust. (Unless using `[[no_unique_address]]` in C++.)

> - The field is first `pad_to_align`. Then, the field is added to the struct with alignment decreased to M. The packing requirement overrides the alignment requirement. (GCC, `#[repr(Rust)]`, `#[repr(C)]` on gnu targets, `#[repr(system)]` on non-windows targets)

Reviewer: I think I might understand where the disconnect is coming from. `MyStruct` *does not* have the field of type `uint8_t` pushed into its layout; it has a field of type `MyStructInner` pushed into its layout, and that `MyStructInner` has the full size of 4 bytes.

No additional alignment needs to be applied to push the field, because the `MyStructInner` type is already sufficiently sized and aligned for the `MyStructInner` value.

> - The field is first `pad_to_align`. Then, the field is added to the struct with alignment decreased to M. The packing requirement overrides the alignment requirement. (GCC, `#[repr(Rust)]`, `#[repr(C)]` on gnu targets, `#[repr(system)]` on non-windows targets)

Reviewer 3: I don't even know what "the field is first `pad_to_align`" even means. First of all I can't quite parse it from a grammatical PoV, and secondly the output of a layout algorithm is the offset of each field of the type (that is, the immediate fields listed in the type declaration, not the recursive ones). So it's completely unclear what this business with recursive fields here is supposed to mean. It sounds like it is talking about the offset of a field of an inner type in the outer type (like, the offset of a field of `MyStructInner` in `MyStruct`), but that makes no sense.

> - The field is first `pad_to_align`. Then, the field is added to the struct with alignment decreased to M. The packing requirement overrides the alignment requirement. (GCC, `#[repr(Rust)]`, `#[repr(C)]` on gnu targets, `#[repr(system)]` on non-windows targets)

Author: Ok I see where the disconnect came from as well. I was under the impression that `MyStructInner` would have a size of 1 byte and an alignment of 4 bytes. If `MyStructInner` already have a size of 4 bytes, `pad_to_align` is indeed not necessary, although it wouldn't be incorrect.

> - The field is added to the struct with alignment increased to N. The alignment requirement overrides the packing requirement. (MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets)

Author: We're not introducing a new type of alignment. Conceptually:

- `packed(N)` changes the alignment of the inner fields to be `N`.
- `aligned(M)` changes the alignment of the entire structure to be `M`.

When both rules apply to the same type, (that's when a packed type transitively contains an aligned field), we have ambiguity, and we need some rules to determine who wins.

So really we're just changing the value returned by `offset_of`. `align_of` will continue to return the base alignment of a type like before.

> - The field is added to the struct with alignment increased to N. The alignment requirement overrides the packing requirement. (MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets)

Reviewer: But it is a different kind of alignment. `i64` has an alignment of 8, which can be lowered by packed. `#[repr(align(8))] i64` has an alignment of 8, which cannot be lowered by packed. `#[repr(align(4))] i64` has an alignment of 8, which can be lowered to 4 by packed. What is this if not a new kind of alignment? Attributes apply to the type they decorate, not to all types which transitively contain the decorated type.

> - The field is added to the struct with alignment increased to N. The alignment requirement overrides the packing requirement. (MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets)

Reviewer: To reiterate this point: you end up with "primitive alignment" which can be suppressed by `packed`, originating from primitives; as well as "user alignment" which cannot be overridden by `packed`, originating from source attributes.

Types are not constructed as a list of primitive fields flattened from the relevant types, even as much as the MSVC layout algorithm sometimes pretends that this is the case. Types are instead defined compositionally, and a field of a struct or enum type is no different from that of some primitive type.

> - The field is added to the struct with alignment increased to N. The alignment requirement overrides the packing requirement. (MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets)

Reviewer 3: And here "user alignment" would be what is ultimately returned by `align_of`, i.e. the alignment required for references to this type?

This reminds me a lot of the discussions around `u64` on 32bit MSVC, which requires separating "alignment of references to this type" and "alignment of struct fields to this type"... is the new kind of alignment proposed in this RFC the same as as that, or does MSVC actually have 3 kinds of alignment?

> - The field is added to the struct with alignment increased to N. The alignment requirement overrides the packing requirement. (MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets)

Author: Well I think the "alignment lowerbility" is only a concept relevant to the specific case of nested aligned types in #[repr(C)] structs; introducing a brand new user-queryable `std::mem::align_attribute` seems a bit overkill. The only thing that the user might be interested in would be the offset of the corresponding field within the packed structure. Whether the alignment can be overriden by `packed` is an implementation detail of `packed` that the user should not have to care about.

> - The field is added to the struct with alignment increased to N. The alignment requirement overrides the packing requirement. (MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets)

Reviewer 3: Whether that new kind of alignment can be queried by the user is a secondary question. But if this concept exists it needs to be named and discussed in the RFC and the Rust Reference, and will become part of the models of Rust that formally describe its behavior.

This overlaps with my [comment over here](https://github.com/rust-lang/rfcs/pull/3718#discussion_r1824936376) saying that the algorithm as you currently describe it is quite hard to understand.

Also `align_attribute` seems like an odd name, it's not the attribute that you need here. It's the fact that types now have two kinds of alignment, that need to have distinct names and that are used in distinct situations. I don't fully understand it yet, that's why I'd like you to explain it in the RFC. :D

> - The field is added to the struct with alignment increased to N. The alignment requirement overrides the packing requirement. (MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets)

Reviewer 4: I think any new kind of alignment should be exposed to user code specifically because user code may want to have its own layout algorithm (e.g. for JIT). Maybe add it as a new field to `Layout` and add methods for msvc struct layout?

> - The field is added to the struct with alignment increased to N. The alignment requirement overrides the packing requirement. (MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets)

Reviewer: Adding fields to `core::alloc::Layout` will be a difficult sell, because `Layout` is part of the global dynamic allocation ABI, and adding more fields to allocation queries which are not needed for the actual allocation wouldn't be very zero cost abstraction of us.

> - The field is added to the struct with alignment increased to N. The alignment requirement overrides the packing requirement. (MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets)

Reviewer 4: actually, now that I'm thinking about it, you'd probably want `repr(MS)` to be available on non-windows too since you'd need it in writing software like [Wine](https://www.winehq.org/) where you have to interface with both the host system and with Windows programs running inside your Win32 API implementation

> - The field is added to the struct with alignment increased to N. The alignment requirement overrides the packing requirement. (MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets)

Reviewer 5: > actually, now that I'm thinking about it, you'd probably want `repr(MS)`

Hmm, GCC does have [`ms_abi`/`sysv_abi` function `__attribute__`s](https://gcc.gnu.org/onlinedocs/gcc/x86-Function-Attributes.html#index-ms_005fabi-function-attribute_002c-x86) for such applications ... and also [`ms_struct`/`gcc_struct` type `__attribute__`s](https://gcc.gnu.org/onlinedocs/gcc/x86-Type-Attributes.html#index-ms_005fstruct-type-attribute_002c-x86), for that exact purpose.

Though I guess GCC doesn't do this perfectly, or clang wouldn't have different ABI between `*-windows-msvc` and `*-windows-gnu` (except I guess the dependency on MS CRT vs. libgcc for those little assembly-code helpers to do various arithmetic and such, and any differences due to the different headers, of course).

(Clang doesn't make any attempt to support `gcc_struct`; in fact, `gcc_struct` does not appear anywhere in the llvm source tree at the moment.)

- Feature Name: `layout_packed_aligned`
- Start Date: 2024-10-24
- RFC PR: [rust-lang/rfcs#3718](https://github.com/rust-lang/rfcs/pull/3718)
- Rust Issue: [rust-lang/rust#100743](https://github.com/rust-lang/rust/issues/100743)

# Summary
[summary]: #summary

This RFC makes it legal to have `#[repr(C)]` structs that are:
- Both packed and aligned.
- Packed, and transitively contains`#[repr(align)]` types.

> It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.

Reviewer: No, the only difference between `#[repr(C)]` and `#[repr(system)]` by this RFC is that on pc-windows-gnu `#[repr(system)]` is the MSVC layout while `#[repr(C)]` is the GCC layout. On all other targets `#[repr(system)]` and `#[repr(C)]` are identical (per this RFC).

> It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.

Reviewer 4: I think `repr(linear)` is independently useful and we should have all 3. that said, I am concerned that current `repr(C)` code often means "I want stable linear layout" rather than "I want whatever weirdness the C compiler decides to use", so I think deprecating `repr(C)` and replacing it with `repr(linear)`, `repr(bikeshed_other_C)` and `repr(system)` is worth considering.

> It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.

Reviewer 3: Indeed, this RFC proposes a *different* distinction between `repr(C)` and `repr(system)` than what has been previously discussed in other threads.

Since the distinction between the two layouts we have here is Windows-only, I wonder if it should be some Windows-only name, like `repr(msvc)` or so? Is there a good reason to even make both of them available on all targets -- effectively exporting a Windows-only complication to other, saner platforms?

> It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.

Reviewer: > Is there a good reason to even make both of them available on all targets

The reason is that this is also how the `extern "system"` function ABI string works. Code uses `extern "system"` to link to libraries which use `__stdcall` on Windows platforms, and in the same way code would use `#[repr(system)]` for linking to a dylib which provides builds with only the MSVC toolchain for Windows targets.

I don't necessarily endorse this option, but it is logically consistent with how Rust already uses `"system"` as an ABI string.

> It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.

Author: I do think that having `repr(linear)` or some kind of stable ABI valid for all platforms will be a good idea, although it's out-of-scope for this RFC. Additionally, this RFC will not impact those people already using `#[repr(C)]` for linear layouts, since it does not modify the behavior when you do not have `aligned` types nested within `packed` types, which was previously forbidden.

> It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.

Reviewer 3: It was previously forbidden but the check has [some gaps](https://github.com/rust-lang/rust/issues/100743#issuecomment-1229343705), so this *can* impact existing code.

It also breaks existing code that assumes that all `repr(C)` types are laid out according to the rules described [here](https://doc.rust-lang.org/reference/type-layout.html#the-c-representation).

> It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.

Reviewer 3: > The reason is that this is also how the extern "system" function ABI string works. Code uses extern "system" to link to libraries which use __stdcall on Windows platforms, and in the same way code would use #[repr(system)] for linking to a dylib which provides builds with only the MSVC toolchain for Windows targets.

From what I understand, `extern "system"` is the same on MSVC and GNU Windows targets? So this is IMO a false analogy then, making it more confusing than if we instead use a name that more explicitly represents that Windows has two ABIs, which we support with two target triples, and you might want to write code that talks with the "other" ABI.

Speaking of which, how would a program for the MSVC target lay out its type in the right way to call a GNU ABI library? That does not seem possible with this proposal.

> It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.

Reviewer: Yes, `extern "system"` is `extern "stdcall"` on all Windows targets. Just like how this RFC makes `repr(C)` on both windows-msvc and windows-gnu behave as `repr(MS)`. The difference is `repr(C)`, which switches between msvc/gnu respectively.

> It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.

Reviewer: Oops, yes, that's correct.

> It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.

Reviewer 6: I don't think we should introduce `system` here. I'd be happy to have the RFC making aligned/packed work together, and we could *potentially* introduce a more specific name for the unique case needed on windows-gnu targets if that's needed, but I don't think we should call that "system" here.

> It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.

Reviewer 3: We also had similar [trouble with AIX](https://internals.rust-lang.org/t/repr-c-aix-struct-alignment/21594), though it may seem like that can in the end be resolved without changing the layout algorithm... it's not entirely clear yet.

But more generally, we could totally encounter a target that specifies its C structs to have some funny layout rules, e.g. requiring certain things to be over-aligned. And for `repr(C)` on ZST, there's really nothing sensible we can do since the C ABI cannot express such types. AFAIK C compilers will typically ensure every struct has size at least 1; should we also adopt that rule and if not, why not?

> It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.

Reviewer 7: > I don't think we should introduce `repr(system)` here. A general mechanism with a generic name like that doesn't seem appropriate for something needed _only_ on the `windows-gnu` targets for using the MSVC ABI.

Do we know *for certain* that we will never get another target like this? As in, a target meant to interoperate with a non-standard C compiler for the system, with a different ABI than the standard system C compiler.

> It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.

Reviewer 8: Suppose that we did. Would it actually be useful to have a "system" repr that could be used for both cases? I would expect most cases of such a repr would be closely tied to the OS.

#[repr(C, packed(2), align(4))]
struct Foo { // Alignment = 4, Size = 8
    a: u8,   // Offset = 0
    b: u32,  // Offset = 2
}
```

This is commonly needed when Rust is being used to interop with existing C and C++ code bases, which may contain
unaligned types. For example in `clang` it is possible to create the following type definition, and there is
currently no easy way to create a matching Rust type:

```cpp
struct  __attribute__((packed, aligned(4))) MyStruct {
    uint8_t a;
    uint32_t b;
};
```

Currently, `#[repr(packed(_))]` structs cannot be `#[repr(align(_))]` or transitively contain `#[repr(align(_))]` types. Attempting to do so results in a [hard error](https://doc.rust-lang.org/nightly/error_codes/E0588.html).

This behavior was added in the [original implementation](https://github.com/rust-lang/rust/issues/33158) of `#[repr(packed)]` due to concerns over differing behavior between msvc and gcc/clang. This makes it cumbersome or even impossible to produce C-compatible struct layouts in Rust when the corresponding C types were annotated with both `packed` and `aligned`.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## `#[repr(C)]`

> When `align` and `packed` attributes exist on the same type, or when `packed` structs transitively contains `align` types,

Reviewer 8: Is there any difference in behavior between the c, system, and rust for types with both packed and align?

My expectation would be that regardless of whether it is c, system, or rust, if I have `repr(packed(2),align(4))`  the alignment of the overall type is 4, and the alignment of it's fields is at most 2 unless maybe one of those fields itself has an alignment specified. My reading of the reference section agrees with that, but the guide section is a little ambiguous. 

And FWIW, it would be more intuitive to me if packed ignored the alignment of field types, but had a way to specify higher alignment for individual fields.

Reviewer: As currently written, `#[repr(align)]` used in a field's type definition (or transitively in its fields' types') could still raise the alignment of the type beyond the minimum alignment provided by the attribute, if the "infectious" alignment is in use (MSVC layout).

In MSVC C, such would theoretically be a compilation error (`alignas` cannot be used to lower alignment). I don't know what the behavior of templated C++ is, actually.

This is consistent with the behavior of `repr(align)` for an alignment less than the alignment required for primitive field alignment.

the resulting layout matches the target toolchain ABI.

For example, given:
```c
#[repr(C, align(4))]
struct Foo(u8);
#[repr(C, packed(1))]
struct Bar(Foo);
```

> `align_of::<Bar>()` would be 4 for `*-pc-windows-msvc` and 1 for everything else.

Reviewer 9: IMO this is extremely unintuitive behavior, especially for those of us writing custom derives (like [zerocopy](https://crates.io/crates/zerocopy)). We can't just rely on querying `align_of` - we need to be able to parse a type's definition and understand the semantics of any `repr` attributes on it.

Also, while this is technically not a breaking change, I wouldn't be surprised if some custom derive code implicitly assumes that this behavior isn't possible, and would become unsound in the face of this change. For example, as it stands today, it is valid to assume that a `#[repr(align(N))]` type has alignment at least `N`, but that would stop being valid with this change.

I'd propose that if system-specific behavior like this is required, it'd be better to do it behind a new repr so that the behavior of `repr(C)` remains straightforward.

Reviewer: > as it stands today, it is valid to assume that a `#[repr(align(N))]` type has alignment at least `N`, but that would stop being valid with this change

Did you say this wrong somehow? By this RFC, `align(N)` types are still always aligned to at least `N` when behind a reference. In fact, they're aligned to `N` in more cases, as `packed` types can't underalign them with MSVC layout rules.

What changes is that `packed(M)` types no longer have an alignment of exactly `M`, but instead have an alignment of at least `M`.

> custom derive code might be unsound

Such custom derive code is already likely quite iffy, since `#[derive(Trait)] #[proc_macro] struct` will have the derive see the token stream before the proc macro processes it and can modify all the decorated code arbitrarily. `zerocopy` is aware of this and doesn't permit any attribute it doesn't know the reserved semantics of to be in the type definition, but that's a high, difficult, and restrictive bar to clear most of the time.

Reviewer 9: > > as it stands today, it is valid to assume that a `#[repr(align(N))]` type has alignment at least `N`, but that would stop being valid with this change
> 
> Did you say this wrong somehow? By this RFC, `align(N)` types are still always aligned to at least `N` when behind a reference. In fact, they're aligned to `N` in more cases, as `packed` types can't underalign them with MSVC layout rules.
> 
> What changes is that `packed(M)` types no longer have an alignment of exactly `M`, but instead have an alignment of at least `M`.

You're right, I misread the text. Still, there's a concern: Our derives currently take `#[repr(C, packed(N))]` as a guarantee that the decorated type cannot have alignment greater than `N`. This is consistent with [the Reference](https://doc.rust-lang.org/reference/type-layout.html#the-alignment-modifiers):

> For `packed`, ... the alignments of each field, for the purpose of positioning fields, is the smaller of the specified alignment and the alignment of the field’s type.

IIUC, to make `packed` no longer provide this guarantee is a breaking change with respect to the Reference.

---

There's another concern as well: This makes "whether or not the type has an alignment repr (`packed`/`align`)" part of the type's layout. The Reference currently specifies how types are laid out purely in terms of the sizes and alignments of their field types, and not in terms of any other facts about the field types. With this change, the following two `Bar` types would have different layouts despite containing `Foo` types with identical sizes and alignments:

```rust
#[repr(C, align(4))]
struct Foo(u8);
#[repr(C, packed(1))]
struct Bar(Foo);

#[repr(C)]
struct Foo(u8, [u32; 0]);
#[repr(C, packed(1))]
struct Bar(Foo);
```

Reviewer 2: > ```rust
> #[repr(C, align(4))]
> struct Foo(u8);
> #[repr(C, packed(1))]
> struct Bar(Foo);
> ```

This does not compile today (reference said "a `packed` type cannot transitively contain another `aligned` type"), so we are free to choose whether the two `Bar`s should have same or different layouts after this RFC.

```
error[E0588]: packed type cannot transitively contain a `#[repr(align)]` type
 --> src/lib.rs:4:1
  |
4 | struct Bar(Foo);
  | ^^^^^^^^^^
  |
note: `Foo` has a `#[repr(align)]` attribute
 --> src/lib.rs:2:1
  |
2 | struct Foo(u8);
  | ^^^^^^^^^^
```

Reviewer 9: My point isn't that this is a breaking change, but rather than it breaks the current *mental model*, which is that a type's layout is determined solely by the size and alignment of its fields. This RFC changes that to say that a type's layout is determined by the size, alignment, *and representation* of its fields. Since representation is a fairly complicated concept (it can include repr(C), repr(Int), repr(packed), repr(align), repr(transparent), and some-but-not-all combinations of these), IMO it significantly complicates the mental model of type layout to say that a type's layout also depends upon its fields' representations.

Reviewer 3: Unfortunately, the existing. mental model turned out to be in contradiction to reality, at least on MSVC targets. When you build a model and then reality demonstrates the incorrectness of the model, sometimes the best option is to fix your model. Yes we documented this model in a bunch of places, but... what else could we do? Stick our head into the sand and continue pretending that the simpler model is "good enough"?

So now the layout of a type is determined by the size, alignment, and [*explicitly requested alignment*](https://github.com/rust-lang/rfcs/pull/3718#discussion_r1826095670) of its fields.

Reviewer: > Unfortunately, the existing mental model turned out to be in contradiction to reality, at least on MSVC targets. When you build a model and then reality demonstrates the incorrectness of the model, sometimes the best option is to fix your model.

There is another alternative: argue that because `__declspec(align)` and `#pragma pack` are nonstandard C extensions, we don't necessarily need to be layout-compatible with the system C in these cases. That's basically the argument for a struct with one zero-length array member on MSVC being zero-sized and not align-sized.

Counter to my own point, though, is that `struct alignas(N) Name` *is* standard C*++*, and the MSVC behavior is the same as with `declspec` in C mode. `#pragma pack` is still nonstandard (and I find MSVC's definition of it to be bonkers) but standard type alignas significantly weakens the argument that it isn't necessary to support. 

But if there exists a type on MSVC with basic alignment higher than the default member packing level without also having an explicitly requested alignment (thus not being sufficiently aligned by default), then I would swing towards writing of MSVC layout edge cases as broken.

> This does not compile today

Unfortunately,​ this error is [straightforward to sidestep with generics.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d713f3ab07be7acd453da00eff0d18a7)

```rust
#[repr(align(4))]
struct Aligned;

#[repr(packed(1))]
struct Packed<T = Aligned>(T);

fn main() {
    dbg!(align_of::<Packed>());
    // [src/main.rs:8:5] align_of::<Packed>() = 1
}
```

Reviewer 9: Isn't it the case that all of these issues could be avoided by introducing a new repr rather than changing the meaning of `repr(C)`? I understand that, historically, `repr(C)` was intended to mean "does the same thing as the C compiler", but it's since come to be used for *much* more beyond that, and its current semantics are relied upon by a huge amount of `unsafe` code that would become silently unsound in the face of the changes proposed here. Is there a reason that introducing a new repr (rather than changing the meaning of `repr(C)`) wouldn't be the obvious solution to address all of the present concerns?

Reviewer: The counter argument is that a similarly huge amount of `unsafe` code exists which is relying on `repr(C)` meaning compatible with the target C compiler's structure layout. It comes down to which use case you consider more important.

This RFC's position is that `repr(C)` matching C is the more important property, and I posit that there's more generated C bindings that wouldn't get migrated to a new repr than there is relying on the exact current behavior; the latter usage is much more likely to be actively maintained and tracking this kind of development, whereas generated bindings are perceived more stable.

My personal position is that `repr(C)` matching the target C compiler is only relevant limited to standard C compliant structure layout, but I do see the benefits of both sides of the debate. It's unfortunately similar to splitting size_t/uintptr_t in that Rust assumes the two usages are fully identical, but edge cases mean that that aren't.

alignments were decreased to `M`. However, in general Rust is free to reorder
these fields for optimization purposes, and the only guarantee is that the fields will maintain a minimum alignment of `M`.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

In the following paragraphs, "Decreasing M to N" means:
```
if M > N {
    M = n
}
```

"Increasing M to N" means:
```
if M < N {
    M = N
}
```

`#[repr(align(N))]` increases the base alignment of a type to be N.

`#[repr(packed(M))]` decreases the alignment of the struct fields to be M. Because the base alignment of the type
is defined as the maximum of the alignment for any fields, this also has the indirect result of decreasing the base
alignment of the type to be M.

When the align and packed modifiers are applied on the same type as `#[repr(align(N), packed(M))]`,
the alignment of the struct fields are decreased to be M. Then, the base alignment of the type is
increased to be N.

When a `#[repr(packed(M))]` struct transitively contains a field with `#[repr(align(N))]` type, depending on the

> target triplet, either:

Reviewer 3: This should IMO be rephrased as an algorithm that works one "layer" at a time. Computing the layout of a type T should only consider the fields of T and their properties. It should never recurse into the fields of T.

I don't currently actually understand the proposed spec here, and this should make it a lot easier to understand.

Reviewer 3: My current understanding of the algorithm is as follows:

---

We equip types with a new property, the "explicitly requested alignment". For all base types, this alignment is 1. For structs, it is by default the maximum of the explicitly requested alignments of all fields. For a struct with the `#[repr(align(N))]` attribute, the explicitly requested alignment the maximum of N and the explicitly requested alignments of its fields.

Note that the explicitly requested alignment of a type can never be bigger than the required alignment of the type.

When computing the layout of a packed(P) struct, then currently we ensure each field is aligned to `min(A, P)`, where `A` is the (regular) alignment of the field type. Under the new rules, *only on MSVC targets*, we instead ensure the field is aligned to `max(E, min(A, P))`, where `E` is the explicitly requested alignment. (Due to the aforementioned inequality, this is equivalent to `min(max(E, P), A)`.) This is the only time the explicitly requested alignment of a type has any effect.

---

I am not fully confident that this is correct. Here's a corner case:
```rust
#[repr(C, align(4))]
struct Align4(i32);

#[repr(C, align(2))]
struct Align2(Align4);

#[repr(C, packed)]
struct Packed(Align2);
```
What is the resulting alignment of `Packed` on MSVC? My proposed algorithm says 4.  Is that correct?

Reviewer: Note: `alignas` on the `struct` is allowed in standard C++ but not standard C. It's a common extension to allow using `alignas` in this position in C anyway. [[cppeference]](https://en.cppreference.com/w/c/language/_Alignas)

```c
#include <stdalign.h>
#include <stddef.h>
#incluse <stdint.h>
#include <stdio.h>

struct alignas(4) Align4
{
    int32_t _0;
};

struct alignas(2) Align2
{
    Align4 _0;
};

#pragma pack(push, 1)
struct Packed
{
    Align2 _0;
};
#pragma pack(pop)

struct Group
{
    uint8_t _0;
    Align4 _1;
};

#pragma pack(push, 1)
struct P
{
    uint16_t _0;
    Group _1;
};
#pragma pack(pop)

int main()
{
    printf("alignof(Packed) = %zu\n", alignof(Packed));
    printf("\n");
    printf("offsetof(P, _0) = %zu\n", offsetof(P, _0));
    printf("offsetof(P, _1) = %zu\n", offsetof(P, _1));
    printf("offsetof(P, _1._0) = %zu\n", offsetof(P, _1) + offsetof(Group, _0));
    printf("offsetof(P, _1._1) = %zu\n", offsetof(P, _1) + offsetof(Group, _1));
}
```

(I wrote this from mobile, so I'm moving to my desktop to actually test this now. IIUC, `Align2` should be an error on MSVC, and you have the layouts otherwise correct. Time to test…)

Reviewer 3: Oh, so having an `align` attribute on a struct where a field already has a higher explicitly requested alignment is an error? Should this also be an error in Rust? The RFC doesn't say so, and it would be a breaking change... but that should at least be mentioned in the RFC. It might be worth a warning if someone writes a `repr(C)` type in Rust that couldn't be written in C.

Reviewer: https://c.godbolt.org/z/es1cGPhz9

On MSVC 19.40 (VS 17.10) in C mode,

```c
#include <stdalign.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>

__declspec(align(4))
struct Align4
{
    int32_t _0;
};

__declspec(align(2))
struct Align2
{
    struct Align4 _0;
};

#pragma pack(push, 1)
struct Packed
{
    struct Align2 _0;
};
#pragma pack(pop)

struct Group
{
    uint8_t _0;
    struct Align4 _1;
};

#pragma pack(push, 1)
struct P
{
    uint16_t _0;
    struct Group _1;
};
#pragma pack(pop)

int main()
{
    printf("alignof(Packed) = %zu\n", alignof(struct Packed));
    printf("\n");
    printf("offsetof(P, _0) = %zu\n", offsetof(struct P, _0));
    printf("offsetof(P, _1) = %zu\n", offsetof(struct P, _1));
    printf("offsetof(P, _1._0) = %zu\n", offsetof(struct P, _1) + offsetof(struct Group, _0));
    printf("offsetof(P, _1._1) = %zu\n", offsetof(struct P, _1) + offsetof(struct Group, _1));
}
```

gives

```text
alignof(Packed) = 4

offsetof(P, _0) = 0
offsetof(P, _1) = 4
offsetof(P, _1._0) = 4
offsetof(P, _1._1) = 8
```

Using `alignas`/`_Alignas` on the type (`alignas(N) struct Tag`) instead of `__declspec(align)` gives

> warning C5274: behavior change: _Alignas no longer applies to the type 'Align4' (only applies to declared data objects)

and results of 1 / 0, 2, 2, 6; fully just ignoring the `alignas` modifier. Note that standard C does not permit the use of `alignas` for struct definitions. C++ does. In C++ mode (`/std:c++latest`, using `struct alignas(N) Tag`), MSVC gives:

```text
alignof(Packed) = 4

offsetof(P, _0) = 0
offsetof(P, _1) = 4
offsetof(P, _1._0) = 4
offsetof(P, _1._1) = 8
```

along with a warning:

> warning C4359: 'Align2': Alignment specifier is less than actual alignment (4), and will be ignored.

Reviewer: For completeness, clang *does not* seem to implement this fully, unless I made a mistake: [[godbolt]](https://c.godbolt.org/z/rTGdazbxE)

<details>

```c
#include <stdalign.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>

struct [[gnu::ms_struct]] Align4
{
    alignas(4)
    int32_t _0;
};

struct [[gnu::ms_struct]] Align2
{
    // alignas(2) // error: requested alignment is less than minimum alignment of 4 for type 'struct Align4'
    struct Align4 _0;
};

struct [[gnu::packed]] Packed
{
    struct Align2 _0;
};

struct Group
{
    uint8_t _0;
    struct Align4 _1;
};

struct [[gnu::packed]] P
{
    uint16_t _0;
    struct Group _1;
};

int main()
{
    printf("alignof(Packed) = %zu\n", alignof(struct Packed));
    printf("\n");
    printf("offsetof(P, _0) = %zu\n", offsetof(struct P, _0));
    printf("offsetof(P, _1) = %zu\n", offsetof(struct P, _1));
    printf("offsetof(P, _1._0) = %zu\n", offsetof(struct P, _1) + offsetof(struct Group, _0));
    printf("offsetof(P, _1._1) = %zu\n", offsetof(struct P, _1) + offsetof(struct Group, _1));
}
```

```text
alignof(Packed) = 1

offsetof(P, _0) = 0
offsetof(P, _1) = 2
offsetof(P, _1._0) = 2
offsetof(P, _1._1) = 6
```

</details>

> - The field is first `pad_to_align`. Then, the field is added to the struct with alignment decreased to M. The packing requirement overrides the alignment requirement. (GCC, `#[repr(Rust)]`, `#[repr(C)]` on gnu targets, `#[repr(system)]` on non-windows targets), or

Reviewer 3: It seems like the parenthetical here is meant to say when which case applies? That's quite confusing, please state the condition *before* the consequence -- just like in code, `if condition { ... } else { ... }`.

Author: Fixed. Thanks!

> - The field is first `pad_to_align`. Then, the field is added to the struct with alignment decreased to M. The packing requirement overrides the alignment requirement. (GCC, `#[repr(Rust)]`, `#[repr(C)]` on gnu targets, `#[repr(system)]` on non-windows targets), or
> - The field is added to the struct with alignment increased to N. The alignment requirement overrides the packing requirement. (MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets)

Reviewer: Minor wording improvement that I believe communicates your intent better without any major substantive change.

```suggestion
- The field is added to the struct with alignment decreased to M. The packing requirement overrides the alignment requirement. (This is the case for GCC, `#[repr(Rust)]`, `#[repr(C)]` on gnu targets, and `#[repr(system)]` on non-windows targets.)
- The field is added to the struct with alignment decreased to M and then increased to N. The alignment requirement overrides the packing requirement. (This is the case for MSVC, `#[repr(C)]` on msvc targets, `#[repr(system)]` on windows targets.)
```

Author: Thank you!

# Drawbacks

> [drawbacks]: #drawbacks

Reviewer 3: Due to [this](https://github.com/rust-lang/rust/issues/100743#issuecomment-1229343705), this RFC will actually *change* the layout of some types that are currently accepted on stable, on MSVC targets. That should be discussed as a drawback.

> Historically the meaning of `#[repr(C)]` has been somewhat ambiguous. When someone puts `#[repr(C)]` on their struct, their intention could be one of three things:

Reviewer 3: https://doc.rust-lang.org/reference/type-layout.html#the-c-representation actually documents the meaning of `repr(C)` quite clearly: it means types are laid out linearly, according to a fixed algorithm. So this RFC is proposing a breaking change, and unsafe code that relies on what is documented in the Reference might become subtly unsound if this RFC gets implemented.

That should at least be mentioned and discussed as a drawback.

Reviewer 9: Isn't such a breaking change an explicit violation of Rust's stability policy? Though I can't actually find where it's documented ATM, my understanding is that breaking changes are only permitted in the following cases:
- The breakage is due to type inference
- The breakage is required in order to fix a security issue

Reviewer 9: As described in https://github.com/rust-lang/rfcs/pull/3718#discussion_r1825075502, it should be simple to avoid this problem by introducing a new repr (in addition to the proposed `#[repr(system)]`) instead of changing the behavior of `#[repr(C)]`.

Reviewer 3: It's definitely in a gray area. One cold also argue that the existing `repr(C)` is simply wrong/unsound, and we have to fix it to satisfy its promise of providing type layout compatible with the current target's C ABI.

Reviewer 4: iirc someone suggested deprecating `repr(C)` entirely and replacing it with `repr(linear)`, `repr(really_C_bikeshed)`, and other `repr`s as necessary.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

This RFC clarifies that:
- `repr(C)` must interoperate with the C compiler for the target.
- `repr(system)` must interoperate with the operating system APIs for the target.
- Similiar to Clang, `repr(C)` does not guarantee consistent layout between targets.

Alternatively, we can also create syntax that allows the user to specify exactly which semantic to use when packed structs transitively contains aligned fields.
For example, a new attribute: #[repr(align_override_packed(N))] that can be used when the behavior of the child overriding the parent alignment is desired.

#[repr(align(N))] #[repr(packed)] can be used together to get the opposite behavior, parent/outer alignment wins.

Explicitly specifying the pack/align semantic has the drawback of complicating FFI. For example, you might need two different definition files depending on the target.

Therefore, a stable layout across compilation target should be relegated as future work.

# Prior art
[prior-art]: #prior-art

Clang matches the Windows ABI for `x86_64-pc-windows-msvc` and matches the GCC ABI for `x86_64-pc-windows-gnu`.

> MinGW always uses the GCC ABI.

Reviewer 3: Is there prior art for a compiler that can lay out types both using the Windows ABI and the GCC ABI for code within a single target? If yes, how are they distinguishing the two? If no, why does Rust need this ability?

Reviewer 4: gcc apparently supports that by using the [`ms_struct` attribute](https://gcc.gnu.org/onlinedocs/gcc/x86-Type-Attributes.html#index-ms_005fstruct-type-attribute_002c-x86)

Reviewer 3: So that would correspond tom in Rust
- have `repr(C)` on win-gnu targets match non-win targets
- have a separate window-only `repr(MS)` to ask for the msvc layout

Reviewer: For a more full parallel to `extern "ABI"`, we should also support `repr(GCC)`. Then `repr(C)` is a sort of alias to `repr(GCC)` or `repr(MS)` chosen by the target, like `extern "C"` is an alias (strongly newtyped) to `"sysv64"`/`"win64"` (etc).

Reviewer: clang [appears to also implement `[[gnu::ms_struct]]`](https://clang.llvm.org/docs/AttributeReference.html#ms-struct) has [an open PR implementing `[[gnu::gcc_struct]]`](https://github.com/llvm/llvm-project/pull/71148)

- Feature Name: `layout_packed_aligned`
- Start Date: 2024-10-24
- RFC PR: [rust-lang/rfcs#3718](https://github.com/rust-lang/rfcs/pull/3718)
- Rust Issue: [rust-lang/rust#100743](https://github.com/rust-lang/rust/issues/100743)

# Summary
[summary]: #summary

This RFC makes it legal to have `#[repr(C)]` structs that are:
- Both packed and aligned.
- Packed, and transitively contains`#[repr(align)]` types.

It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.
It has the same behavior as `#[repr(C)]` except on `*-pc-windows-gnu` targets where it uses the msvc layout
rules instead.

> # Motivation

Reviewer 9: Can we explicitly document that a goal of this RFC is to make sure that existing uses of `#[repr(C)]` are required to clarify their intent (ie, to indicate compatibility with a C compiler or to indicate a "linear" layout)? In particular, this implies that `#[repr(C)]` itself should be deprecated, and the two behaviors should both be supported by new reprs. That will ensure that code doesn't silently do either of the following:
- Change its behavior (if we decide to change `#[repr(C)]` to mean "compatible with the C compiler")
- Remain subtly buggy (if we decide to keep `#[repr(C)]`'s behavior, and that code was intended to be compatible with the C compiler)
